framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -529,49 +530,119 @@ public TypeArgumentInference getTypeArgumentInference() {
 
-        if (typeQualifiersAnnotation == null) {
-            // If not, try the Checker
-            classType = checker.getClass();
+            // First see if the AnnotatedTypeFactory has @TypeQualifiers
+            Class<?> classType = this.getClass();
             typeQualifiersAnnotation = classType.getAnnotation(TypeQualifiers.class);
-        }
 
-        if (typeQualifiersAnnotation != null) {
-            Set<Class<? extends Annotation>> typeQualifiers = new HashSet<Class<? extends Annotation>>();
-            for (Class<? extends Annotation> qualifier : typeQualifiersAnnotation.value()) {
-                typeQualifiers.add(qualifier);
+            if (typeQualifiersAnnotation == null) {
+                // If not, try the Checker
+                classType = checker.getClass();
+                typeQualifiersAnnotation = classType.getAnnotation(TypeQualifiers.class);
+            }
+
+            if (typeQualifiersAnnotation != null) {
+                for (Class<? extends Annotation> qualifier : typeQualifiersAnnotation.value()) {
+                    typeQualifiers.add(qualifier);
+                }
+                return Collections.unmodifiableSet(typeQualifiers);
+            } else {
+                return Collections.emptySet();
             }
-            return Collections.unmodifiableSet(typeQualifiers);
         }
+    }
 
-        return Collections.emptySet();
+    /**
+     * Loads all qualifiers contained in the qual directory of a checker via
+     * reflection, and has the option to include the PolyAll qualifier and a
+     * list of manually stated qualifiers (eg ones found in a different folder
+     * than the checker). The type qualifiers that are automatically loaded must
+     * have the {@link java.lang.annotation.Target Target} meta-annotation with
+     * the value of {@code ElementType.TYPE_USE} and/or
+     * {@code ElementType.TYPE_PARAMETER}, otherwise the loader skips the
+     * qualifier. The ones that do not have the {@code ElementType.TYPE_USE}
+     * and/or {@code ElementType.TYPE_PARAMETER} meta-annotation value must be
+     * manually stated even if they are in the same folder as the checker's qual
+     * folder.
+     * 
+     * This method can be called in the overridden versions of
+     * createSupportedTypeQualifiers() in each checker to load {@link PolyAll},
+     * and/or load a manually stated list of qualifiers.
+     * 
+     * @param excludePolyAll
+     *            setting this to true would skip adding {@link PolyAll} to the
+     *            list of qualifiers
+     * @param manuallyListedQualifiers
+     *            a set of manually listed qualifiers to be added to the
+     *            qualifier set, for example, it is used frequently to add
+     *            Bottom qualifiers
+     * @return a set of annotation class instances
+     */
+    protected Set<Class<? extends Annotation>> loadTypeQualifiersFromQualDir(final boolean excludePolyAll,
```

Comment:
I would make this two methods:
getQualifiersFromQualsDirWithPolyAll(...)
getQualifiersFromQualsDirWithOutPolyAl(..)
They can just call this method with true or false, otherwise, it makes code calling this method harder to read.